File: backend/services/member.js
import _ from "lodash";
import axios from "axios";

import { api } from "../utils/controller-api.js";
import { db } from "../utils/db.js";
import { getZTAddress } from "../utils/zt-address.js";

let ZT_ADDRESS = null;
getZTAddress().then(function (address) {
  ZT_ADDRESS = address;
});

async function getPeer(mid) {
  try {
    const peer = await api.get("peer/" + mid);
    return peer.data;
  } catch (err) {
    return;
  }
}

async function getMemberAdditionalData(data) {
  // DB BUG INITIALIZATION MIGRATION
  const network = db.get("networks").find({ id: data.nwid });
  network.defaults({ members: [] }).get("members").write();
  // END MIGRATION SECTION

  const member = db
    .get("networks")
    .find({ id: data.nwid })
    .get("members")
    .find({ id: data.id });

  const additionalData = member.get("additionalConfig").value() || {};
  const lastOnline = member.get("lastOnline").value() || 0;

  const peer = await getPeer(data.id);
  let peerData = {};
  if (peer && !_.isEmpty(peer)) {
    peerData.latency = peer.latency;
    if (peer.latency !== -1) peerData.online = 1;
    if (peer.latency == -1) peerData.online = 2;
    peerData.clientVersion = peer.version;
    if (peer.paths.length > 0) {
      let path = peer.paths.filter((p) => {
        let ret = p.active && !p.expired;
        if (typeof p.preferred !== "undefined") {
          ret = ret && p.preferred;
        }
        return ret;
      });
      if (path.length > 0) {
        peerData.lastOnline = path[0].lastReceive;
        peerData.physicalAddress = path[0].address.split("/")[0];
        peerData.physicalPort = path[0].address.split("/")[1];
      }
    }
  } else {
    peerData.online = 0;
  }

  delete data.lastAuthorizedCredential;
  delete data.lastAuthorizedCredentialType;
  delete data.objtype;
  delete data.remoteTraceLevel;
  delete data.remoteTraceTarget;

  return {
    id: data.nwid + "-" + data.id,
    clock: new Date().getTime(),
    networkId: data.nwid,
    nodeId: data.id,
    controllerId: ZT_ADDRESS,
    // @ts-ignore
    lastOnline: lastOnline,
    ...additionalData,
    ...peerData,
    config: data,
  };
}

async function filterDeleted(nwid, mid) {
  const member = db
    .get("networks")
    .find({ id: nwid })
    .get("members")
    .find({ id: mid });

  let deleted = member.get("deleted").value() || false;
  if (!deleted) return mid;
  else return;
}

export async function getMembersData(nwid, mids) {
  const prefix = "/controller/network/" + nwid + "/member/";
  const filtered = (
    await Promise.all(mids.map(async (mid) => await filterDeleted(nwid, mid)))
  ).filter((item) => item !== undefined);
  const links = filtered.map((mid) => prefix + mid);

  const multipleRes = await axios
    .all(links.map((l) => api.get(l)))
    .then(
      axios.spread(function (...res) {
        return res;
      })
    )
    .catch(function (err) {
      return [];
    });

  let data = Promise.all(
    multipleRes.map((el) => {
      return getMemberAdditionalData(el.data);
    })
  );

  return data;
}

export async function updateMemberAdditionalData(nwid, mid, data) {
  if (data.config && data.config.authorized) {
    db.get("networks")
      .filter({ id: nwid })
      .map("members")
      .first()
      .filter({ id: mid })
      .first()
      .set("deleted", false)
      .write();
  }

  let additionalData = {};

  if (data.hasOwnProperty("name")) {
    additionalData.name = data.name;
  }
  if (data.hasOwnProperty("description")) {
    additionalData.description = data.description;
  }

  if (additionalData) {
    const member = db
      .get("networks")
      .filter({ id: nwid })
      .map("members")
      .first()
      .filter({ id: mid });
    if (member.value().length) {
      member
        .map("additionalConfig")
        .map((additionalConfig) => _.assign(additionalConfig, additionalData))
        .write();
    } else {
      additionalData = { name: "", description: "" };

      if (data.hasOwnProperty("name")) {
        additionalData.name = data.name;
      }
      if (data.hasOwnProperty("description")) {
        additionalData.description = data.description;
      }
      db.get("networks")
        .filter({ id: nwid })
        .map("members")
        .first()
        .push({ id: mid, additionalConfig: additionalData })
        .write();
    }
  }
}

export async function deleteMemberAdditionalData(nwid, mid) {
  // ZT controller bug
  /* db.get("networks")
       .find({ id: nwid })
       .get("members")
       .remove({ id: mid })
       .write();
  */

  await updateMemberAdditionalData(nwid, mid, {});

  db.get("networks")
    .filter({ id: nwid })
    .map("members")
    .first()
    .filter({ id: mid })
    .first()
    .set("deleted", true)
    .write();
}

File: backend-py/services/member.py
import sys, os, time

import sys, os
sys.path.append(os.getcwd())
from constants.constants import *

from vars import DARWIN, LINUX
sys_vars = DARWIN() if os.uname().sysname == "Darwin" else LINUX()


class MEMBER():
    """Classe um die Anfragen rund ums MemberVerwaltung zu handeln"""
    
    def getPeer(self, mid : str) -> dict:
        
        return zerotier_api.make_request("GET", f"/peer/{mid}")

    def filterDeleted(self, nwid : str, mid : str) -> str:
        try:
            if db.read_networks_member_value(nwid, mid).get("deleted") != True:
                
                return mid
        except:
            
            return mid


    def getMembersData(self, nwid : str , mids : str) -> list:
        members = []
        for mid in mids:
            if self.filterDeleted(nwid, mid):
                member = zerotier_api.make_request("GET", f"/controller/network/{nwid}/member/{mid}").json()
                members.append(self.getMemberAdditionalData(member))
            
        return members

    def getMemberAdditionalData(self, data : dict) -> dict:
        
        #DB BUG INITIALIZATION MIGRATION
        if db.read_networks_member_value(data.get("nwid"),data.get("id")) == None:
            members = db.read_networks_value(data.get("nwid"))[0].get("members")
            if members != None:
                members.append({"id" : data.get("address")})
                db.table_update("networks", {"members" : members } , db.read_networks_value(data.get("nwid"))[0].doc_id)
            else:
                db.table_update("networks", {"members" : [{"id" : data.get("address")}] } , db.read_networks_value(data.get("nwid"))[0].doc_id)
        #END MIGRATION SECTION
        
        
        peerData = {}
        zt = zerotier_api.make_request("GET", "/status").json()["address"]
        peer = self.getPeer(data.get("id")).json()

        if (peer):
            peerData["latency"] = peer.get("latency")
            peerData["online"] = 1 if peer.get("latency") != -1 else 2
            peerData["clientVersion"] = peer.get("version")
            if peer.get("paths"):
                peerData["lastOnline"] = peer.get("paths")[0]["lastReceive"]
                peerData["physicalAddress"] = peer.get("paths")[0]["address"].split("/")[0]
                peerData["physicalPort"] = peer.get("paths")[0]["address"].split("/")[1]
        else:
            peerData["online"] = 0

        del data["lastAuthorizedCredential"]
        del data["lastAuthorizedCredentialType"]
        del data["objtype"]
        del data["remoteTraceLevel"]
        del data["remoteTraceTarget"]
        
        additionalData = {}
        
        try :
            additionalData = db.read_networks_member_value(data.get("nwid"),data.get("id"))["additionalConfig"]
        except:
            additionalData = {}

        memberData = {
            "id"    : data.get("nwid") + "-" + data.get("id"),
            "type"  : "Member",
            "clock" : time.time(),
            "networkId" : data.get("nwid"),
            "nodeId" : data.get("id"),
            "controllerId" : zt,
            **additionalData,
            **peerData,
            "config" : data
        }
        
        return memberData

    def updateMemberAdditionalData(self, nwid : str, mid : str, data : dict) -> None:
        try:
            additionalConfig = db.read_networks_member_value(nwid, mid)["additionalConfig"]
        except:
            additionalConfig = {}
        
        ## Only for Usermanagment
        if type(data) == list:
            data = data[0]
            authorized = data["authorized"] if data["authorized"] == True else False
            db.update_network_members(nwid, mid, {"additionalConfig" : db.merge_two_dicts(additionalConfig, {"authorized" : authorized})} )
        ##
        else:
            if data.get("config") and data.get("config").get("authorized") != None:
                authorized = data.get("config")["authorized"] if data.get("config")["authorized"] == True else False
                db.update_network_members(nwid, mid, {"additionalConfig" : db.merge_two_dicts(additionalConfig, {"authorized" : authorized})} )

            additionalData = data.get("config") if data.get("config") != None else {}

            if (data.get("name")):
                additionalData["name"] = data["name"]

            if (data.get("description")):
                additionalData["description"] = data["description"]

            if additionalData:
                db.update_network_members(nwid, mid, {"additionalConfig" : db.merge_two_dicts(additionalConfig, additionalData)})

    def deleteMemberAdditionalData(self, nwid, mid) -> None:
        db.update_network_members(nwid, mid, {"id": mid, "deleted" : True})